diff --git a/x.c b/x.c
index 210f184..5574a56 100644
--- a/x.c
+++ b/x.c
@@ -2043,6 +2043,9 @@ run:
 	xinit(cols, rows);
 	xsetenv();
 	selinit();
+	if (getenv("ST_FONT") != NULL && getenv("ST_FONT")[0] == '1') { cyclefonts(NULL); }
+	if (getenv("ST_FONTSZ") != NULL) { Arg a; a.f = (float)atoi(getenv("ST_FONTSZ")); zoom(&a); }
+	if (getenv("ST_INVERT") != NULL && getenv("ST_INVERT")[0] == '1') { invert(NULL); }
 	run();
 
 	return 0;
